crates/gitbutler-watcher/src/file_monitor.rs

```diff
@@ -116,28 +118,34 @@ pub fn spawn(
                     tracing::error!(?err, "ignored file watcher error");
                 }
                 Ok(events) => {
-                    let maybe_repo = git::Repository::open(&worktree_path)
-                        .with_context(|| format!("failed to open project repository: {}", worktree_path.display()))
-                        .map(Some)
-                        .unwrap_or_else(|err| {
-                            tracing::error!(
-                                ?err,
-                                "will consider changes to all files as repository couldn't be opened"
-                            );
-                            None
-                        });
-
                     let num_events = events.len();
-                    let classified_file_paths = events
+                    let mut classified_file_paths: Vec<_> = events
                         .into_iter()
                         .filter(|event| is_interesting_kind(event.kind))
                         .flat_map(|event| event.event.paths)
                         .map(|file| {
-                            let kind = maybe_repo
-                                .as_ref()
-                                .map_or(FileKind::Project, |repo| classify_file(repo, &file));
+                            let kind = classify_file(&git_dir, &file);
                             (file, kind)
-                        });
+                        })
+                        .collect();
+                    if classified_file_paths
+                        .iter()
+                        .any(|(_, kind)| *kind == FileKind::Project)
+                    {
+                        if let Ok(repo) = gix::open(&worktree_path) {
+                            if let Ok(index) = repo.index_or_empty() {
+                                if let Ok(mut excludes) = repo.excludes(&index, None, gix::worktree::stack::state::ignore::Source::WorktreeThenIdMappingIfNotSkipped) {
+                                    for (file_path, kind) in classified_file_paths.iter_mut() {
+                                        if let Ok(relative_path) = file_path.strip_prefix(&worktree_path) {
+                                            if excludes.at_path(relative_path, None).map(|platform| platform.is_excluded()).unwrap_or(false) {
```

Comment:
Note that the `None` here means that there is no information about the file type, so it internally treats it like a file. However, some of these paths may be directories, and right now we tend to just pass them along. 

I checked, and there seems to be no quick way to filter them out, as the filesystem events don't provide type information (or I didn't find it).